### DISCUSSION TITLE
Why it always returns only 10 rows of result even if collection.num_entities>1000 and limit=100

### DISCUSSION THREAD

**Discussion by JSilent** [milvus-io/milvus#33678] [Q&A and General discussion] [answered]
Why it always returns only 10 rows of result even if collection.num_entities>1000 and limit=10

The script is as belowed:

collection = Collection(name="test_collection")

collection.load()
print(collection.num_entities)
search_params = {
    "metric_type": "IP",
    "params": {"nprobe": 10}
}

np_query_embeddings = np.array(query_embeddings, dtype=np.float32)
norms = np.linalg.norm([np_query_embeddings], axis=1, keepdims=True)
normalized_query_embeddings = np_query_embeddings / norms

query_vectors = normalized_query_embeddings.tolist()
limit = 1000

results = collection.search(
    data=query_vectors,
    anns_field="embedding",
    param=search_params,
    limit=limit,
    consistency_level="Strong",
)
print(results)

Display:

`8503
["['id: 449617239426872838, distance: 0.7159370183944702, entity: {}', 'id: 449617239426877164, distance: 0.7090992331504822, entity: {}', 'id: 449617239426880450, distance: 0.699631929397583, entity: {}', 'id: 449617239426876554, distance: 0.6782248020172119, entity: {}', 'id: 449617239426876178, distance: 0.6769057512283325, entity: {}', 'id: 449617239426880350, distance: 0.6754627823829651, entity: {}', 'id: 449617239426875291, distance: 0.6698116064071655, entity: {}', 'id: 449617239426877126, distance: 0.6647711396217346, entity: {}', 'id: 449617239426876516, distance: 0.6610612869262695, entity: {}', 'id: 449617239426880437, distance: 0.6571118831634521, entity: {}']"]`

**Comment by JSilent** [comment]
Both version of client and server is 2.2.13.

**Comment by yhmo** [comment]
Possible reason: there are duplicate records using the same ID.

**Reply by JSilent** [reply]
The ID was generated automatically.

`fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
    FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, dim=384)
]`

**Reply by JSilent** [reply]
Using self created ID, the issue still happens.

fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True),
    FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, dim=384)
]
...
mr = collection.insert([[i for i in range(len(normalized_vectors))], normalized_vectors])

`8503
["['id: 221, distance: 0.7159370183944702, entity: {}', 'id: 4547, distance: 0.7090992331504822, entity: {}', 'id: 7833, distance: 0.699631929397583, entity: {}', 'id: 3937, distance: 0.6782248616218567, entity: {}', 'id: 3561, distance: 0.6769057512283325, entity: {}', 'id: 7733, distance: 0.6754627823829651, entity: {}', 'id: 2674, distance: 0.6698116064071655, entity: {}', 'id: 4509, distance: 0.6647711396217346, entity: {}', 'id: 3899, distance: 0.6610612869262695, entity: {}', 'id: 7820, distance: 0.6571118831634521, entity: {}']"]`

**Comment by yhmo** [comment]
Set the "nprobe" to be equal to "nlist" and try again.
Change the index type to be "FLAT" and try again.

**Reply by JSilent** [reply]
Issue still happens.

```
index_params = {
    "metric_type": "IP",
    "index_type": "FLAT",
    "params": {"nlist": 100}
}
```

```
search_params = {
    "metric_type": "IP",
    "params": {"nprobe": 100}
}
```

And the num_entities sometimes is 0.

```
0
["['id: 221, distance: 0.7159370183944702, entity: {}', 'id: 4547, distance: 0.7090992331504822, entity: {}', 'id: 7833, distance: 0.699631929397583, entity: {}', 'id: 3937, distance: 0.6782248616218567, entity: {}', 'id: 3561, distance: 0.6769057512283325, entity: {}', 'id: 7733, distance: 0.6754627823829651, entity: {}', 'id: 2674, distance: 0.6698116064071655, entity: {}', 'id: 4509, distance: 0.6647711396217346, entity: {}', 'id: 3899, distance: 0.6610612869262695, entity: {}', 'id: 7820, distance: 0.6571118831634521, entity: {}']"]
```

**Reply by JSilent** [reply]
```
norms = np.linalg.norm(vectors, axis=1, keepdims=True)
normalized_vectors = vectors / norms
mr = collection.insert([[i for i in range(len(normalized_vectors))], normalized_vectors])
print(mr)
print(collection.num_entities)
```

```
(insert count: 8503, delete count: 0, upsert count: 0, timestamp: 450314665369534468, success count: 8503, err count: 0)
0
```

**Comment by xiaofan-luan** [comment]
1. are u sure the limit the 1000 or 10? you have 100 in title, 10 in description and 1000 in the code. so it's really mess up.
2. We recommend u to use 2.3.17 i you are still under testing.
3.  since you already use flat index, there is no possiblity that index params will affect performance.
My suggestion is:
1. upgrade to 2.3.17
2. recreate a collection 
3. run the script, check if it works.

if not, please offser you test scripts and logs to us.
Thanks for the patience.

**Reply by JSilent** [reply]
Sorry for making you confused. I have tested several values for the limit: 50, 80, 100, 1000. For all of them milvus gave me only 10 rows of result.

I upgraded milvus to 2.4.4 according to your suggestion, but it still returns 10 rows.

I'm uploading the testing notebook. As github doesn't support upload ipynb, please change the extension after you download it.
I did't upload the source pkl file since it contains sensitive information.
[use-milvus.txt](https://github.com/user-attachments/files/15783678/use-milvus.txt)

**Reply by yhmo** [reply, marked as answer]
Use this code to print the results. This code will print the entire results:
```
    for i, result in enumerate(results):
        print("\nSearch result for {}th vector: ".format(i))
        for j, res in enumerate(result):
            print("Top {}: {}".format(j, res))
```

If you use "print(results)", it will always print the first 10 records. Because the search() returns a SearchResult class, and print(results) calls to `__str__()` of this class, and the `__str__()` method always print the first 10 records:
https://github.com/milvus-io/pymilvus/blob/e95f5a5294ff28588c14a2f3f5aae19d9210a47e/pymilvus/client/abstract.py#L517

In my opinion, this is a bad design, brings confusion to users.

**Reply by JSilent** [reply]
With your code it could loop the results. The results have limit number of rows. Thank you both so much for the answers.

```
Search result for 0th vector: 
Top 0: id: 221, distance: 0.7159370183944702, entity: {}
Top 1: id: 4547, distance: 0.7090992331504822, entity: {}
Top 2: id: 7833, distance: 0.699631929397583, entity: {}
...
Top 9: id: 7820, distance: 0.6571118831634521, entity: {}
Top 10: id: 7763, distance: 0.6540083885192871, entity: {}
Top 11: id: 3416, distance: 0.6522840261459351, entity: {}
Top 12: id: 325, distance: 0.6513450741767883, entity: {}
...
Top 96: id: 229, distance: 0.5847136974334717, entity: {}
Top 97: id: 2560, distance: 0.5820827484130859, entity: {}
Top 98: id: 4646, distance: 0.5820591449737549, entity: {}
Top 99: id: 7855, distance: 0.581797182559967, entity: {}
```

**Comment by xiaofan-luan** [comment]
1. how many data are there in test.pkl?
2. if you change limit to 1, how many result did you get?
3. count how many entities are there in the colllection https://milvus.io/docs/get-and-scalar-query.md#Count-entities

**Reply by JSilent** [reply]
1. 8503
2. One row of result.
`data: ["['id: 7872, distance: 0.3031132221221924, entity: {}']"] , cost: 0`
3. 8503
`data: ["{'count(*)': 8503}"] , extra_info: {'cost': 0}`
